core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java

```diff
@@ -428,6 +431,10 @@ public MaterializedConnector(CatalogName catalogName, Connector connector)
             requireNonNull(procedures, format("Connector '%s' returned a null procedures set", catalogName));
             this.procedures = ImmutableSet.copyOf(procedures);
 
+            Set<TableProcedureMetadata> tableProcedures = connector.getTableProcedures();
```

Comment:
There is a decision to be made whether table procedures should be statically or dynamically resolved.

for example, what should be the behavior when a procedure is applicable to certain relation types (eg only tables, or only certain type of tables)?
exposing it as if it existed might feel better and then throw at execute time "this is not applicable", but one could say the same about `$bucket` hidden column, or between-connectors user expectations (users may not  appreciate the fact that hive supports different set of procedure names than iceberg connector).

Also, it might be that we want to overload the name -- expose same procedure _name_, but with different properties or execution mode, or internal implementation. For example "compact table" is _logically_ applicable to transactional and non-transactional tables, but is different operation internally.

Thus, i would suggest we don't register them here, and move the procedure name  resolution from Connector into ConnectorMetadata

Reply:
We had an offline converstation about this one. Let's leave it static for now.
tl;dr is:
* it may be beneficial to be able to resolve a single procedure name to different implementations depending on context (e.g `compact` for Hive transactional table and for non-transactional table can do very different things). Yet we can achieve the same (similar) thing by doing conditional logic within the procedure implementation itself. Actually handling different table types within the procedure can be better, as we may construct nicer (non-generic) exception messages if the procedure is misused. Hence have better UX.
* as for supporting different relation types (tables vs vies vs materialized views) dynamic resolution is just part of the story. The more tricky part is how the rest of SPI should look to support not just tables. It is not straightforward and we do not want to block this PR on resolving that. It should be a follow-up.